### ISSUE TITLE
NHibernate bags are flushed during wcf serialization

### ISSUE THREAD

**Issue by Cyril035** [nhibernate/nhibernate-core#2972]
Hi there,

I used nhibernate v4.1.4 for a long time and all works perfectly until we decide to update to lastest version. (5.3.10) We use it on a wcf service. Data loaded correctly but when objects are serialiazed (with wcf mechanisms) all collections are gone away... :( I precise that all collections load correctly with data inside before serialization.

I try to debug with this snippet

```
var myObjectWithCollections = Session.Query<MyClass>().Single(x => x.Id == id);
var serializer = new NetDataContractSerializer();
var wcfMessage = new FileStream("soapMsg.xml", FileMode.Create);
serializer.WriteObject(wcfMessage , myObjectWithCollections);
```
When serializer.WriteObjectis called, all collection are "flushed" :(

I decide for testing to change the type of one collection from IList<T> to ICollection<T> to test and....tadaaaaaa. It works. Unfortunately, I don't know why and I don't want to change all collection type to this ICollection

After serialization, normally collections are of type PersisentGenericBag which implement ICollection<T> AND IList<T> so it sould work ! I miss something but what ?

NOTE : all my collections are "bags" in my mappings files (hbm.xml). The loading mode was neither eager nor lazy. With 4.1.4 mappings were configured with full loading.

Have you got some ideas cauz I'm a bit confused now :(

Subject posted to [stackoverflow](https://stackoverflow.com/questions/70569750/nhibernate-bags-are-flushed-during-wcf-serialization) too

**Comment by oskarb** [comment]
To clarify, when you say "flushed", do you mean that the collections appear
empty on the client side? Because usually when talking about NHibernate,
"flushing" is about transmitting changes back to the database.

Do they appear empty before server side before serialisation?

**Comment by Cyril035** [comment]
Yeah, sorry for this. I mean collections are filled with data from database and after serialization they are empty

**Comment by bahusoid** [comment]
Does serialization work properly with manually created and populated list?  Something like:
```C#
var newObj =  new MyClass();
newObjc.Collection = new List<Child>();
newObj.Collection.Add(new Child());
`serializer.WriteObject(wcfMessage ,newObj);
```

**Comment by Cyril035** [comment]
@bahusoid 
Yeah, it works. Actually, in session context, NHibernate create a PersisantGenericBag for this collection which is normally serialized. The client should know this type. The wcf fail to serialize this PersistantGenericBag and retrieve a empty generic list instead.

**Comment by bahusoid** [comment]
I think it should be closed as external issue because of:
> I go back to the last working version (4.1.1) and lol... don't work too... Something broken during update.

Or be converted to discussion.

**Comment by Cyril035** [comment]
Hi there,
@bahusoid, even something is going wrong to my project, I don't think it's only the problem. Always this question... why when I change from IList<T> to ICollection<T>, it works perfectly... ?
